test/test_transforms_ad_rules.jl

```diff
@@ -0,0 +1,101 @@
+                msg_new = "$msg: ::$(typeof(x)).$k"
+            else
+                msg_new = "$msg: getfield(::$(typeof(x)), $k)"
+            end
+            EnzymeTestUtils.test_approx(getfield(x, k), getfield(y, k), msg_new; kwargs...)
+        end
+    end
+    return nothing
+end 
+
+@testset "SpeedyTransforms: AD Rules" begin
+
+    @testset "_fourier! Enzyme rules" begin
+        
+        @testset "reverse rule" begin
+            for grid_type in grid_types
+
+                spectral_grid = SpectralGrid(Grid=grid_type)
+                S = SpectralTransform(spectral_grid)
+                grid = rand(spectral_grid.Grid{spectral_grid.NF}, spectral_grid.nlat_half, spectral_grid.nlayers)
+                f_north = S.scratch_memory_north
+                f_south = S.scratch_memory_south
+
+                # not working currenlty, the test is stuck 
+                #test_reverse(SpeedyWeather.SpeedyTransforms._fourier!, Const, (f_north, Duplicated), (f_south, Duplicated), (grid, Duplicated), (S, Const))
+            end
+        end
+    end 
+
+    @testset "Complete Transform Enzyme" begin
+        # make a high level finite difference test of the whole transform
+        # can't use Enzyme or ChainRule Test tools for tests for that
+        for grid_type in grid_types
+
+            spectral_grid = SpectralGrid(Grid=grid_type)
+
+            # forwards 
+            S = SpectralTransform(spectral_grid)
+            dS = deepcopy(S)
+            grid = rand(spectral_grid.Grid{spectral_grid.NF}, spectral_grid.nlat_half, spectral_grid.nlayers)
+            dgrid = zero(grid)
+            specs = zeros(LowerTriangularArray{Complex{spectral_grid.NF}}, spectral_grid.trunc+2, spectral_grid.trunc+1, spectral_grid.nlayers)
+            
+            # seed
+            dspecs = zero(specs)
+            fill!(dspecs, 1+1im)
+
+            autodiff(Reverse, transform!, Const, Duplicated(specs, dspecs), Duplicated(grid, dgrid), Duplicated(S, dS))
+
+            # new seed
+            dspecs2 = zero(specs)
+            fill!(dspecs2, 1+1im)
+
+            # finite difference comparision, seeded with a one adjoint to get the direct gradient
+            fd_jvp = FiniteDifferences.j′vp(central_fdm(5,1), x -> transform(x, S), dspecs2, grid)
+            @test isapprox(dgrid, fd_jvp[1])
+
+            ## now backwards, as the input for spec we use the output of the forward transform
+
+            fill!(dspecs,0)
+            grid = zeros(spectral_grid.Grid{spectral_grid.NF}, spectral_grid.nlat_half, spectral_grid.nlayers)
+            dgrid = similar(grid)
+            fill!(dgrid, 1)
+
+            autodiff(Reverse, transform!, Const, Duplicated(grid, dgrid), Duplicated(specs, dspecs), Duplicated(S, dS))
+
+            # new seed 
+            dgrid2 = similar(grid)
+            fill!(dgrid2, 1)
+
+            fd_jvp = FiniteDifferences.j′vp(central_fdm(5,1), x -> transform(x, S), dgrid2, specs)
+
+            @test isapprox(dspecs, fd_jvp[1])
```

Comment:
We could also test that the roundtrip transform is identity no?

Reply:
What roundtrip exactly? I am not sure I understand. You could differentiate through $\mathcal{S}^{-1}(\mathcal{S}(x))$, but that's just identity, so the derivative wrt the input would just vanish. Or is this exactly what you were thinking of?

Reply:
Vanish=1 no? It's d(S^-1(S(x)))/dx = dx/dx = 1 I thought? And we could test this starting in either space?

Reply:
Ah yes sure, you are right.

Reply:
I added this as tests. 

Starting in spectral space it works very well
Starting in grid space, the result is sort of close to 1, but not quite. Currently I am not sure what goes wrong there.

Reply:
I am not surprised. This depends on what data you start with in grid space, which can hold more information than spectral space, so the first roundtrip is usually lossy. But subsequent ones should be fine. I would start in spectral space, transform to grid space then start the test grid->spectral->grid there

Reply:
We already do that in the tests. But in principal you are right, there's something wrong with the setup for this test. Because the finite difference differentiation yields the same gradient as Enzyme. So it's not an issue of the custom rule.

Reply:
The test passes now, I changed the function to test the identity to not directly overwrite the input data, but instead save the output in a pre-allocated array. This wasn't just an issue though with Enzyme, FiniteDifferences.jl showed the same error. Now it works.